src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,301 @@
+                )?;
+                region.constrain_equal(&config.perm, piece.cell(), cell)?;
+                zs.push(CellValue::new(cell, piece.field_elem()));
+
+                // Assign cumulative sum such that
+                //          z_i = 2^10 * z_{i + 1} + m_{i + 1}
+                // => z_{i + 1} = (z_i - m_{i + 1}) / 2^10
+                //
+                // For a message m = m_1 + 2^K m_2 + ... + 2^{Kl} m_l}, initialize z_0 = m.
+                // We end up with z_{l} = 0.
+                let mut z = piece.field_elem();
+                let inv_2_k = pallas::Base::from_u64(1 << K).invert().unwrap(); // TODO: Constant
+                for (idx, word) in words.iter().enumerate() {
+                    // z_next = (z_cur - m_next) / 2^K
+                    z = z
+                        .zip(*word)
+                        .map(|(z, word)| (z - pallas::Base::from_u64(word as u64)) * inv_2_k);
+                    let cell = region.assign_advice(
+                        || format!("z_{:?}", idx + 1),
+                        config.word,
+                        offset + idx + 1,
+                        || z.ok_or(Error::SynthesisError),
+                    )?;
+                    zs.push(CellValue::new(cell, z))
+                }
+
+                Ok(())
+            };
+
+        // Convert `words` from `Option<Vec<u32>>` to `Vec<Option<u32>>`
+        let words: Vec<Option<u32>> = if let Some(words) = words {
+            words.into_iter().map(Some).collect()
+        } else {
+            vec![None; piece.num_words()]
+        };
+
+        // Decompose message into words using a running sum.
+        decompose_message(region, words)?;
+
+        // Copy in the accumulator x-coordinate
+        let x_a = copy(
+            region,
+            || "Initialize accumulator x-coordinate",
+            config.x_a,
+            offset,
+            &x_a,
+            &config.perm,
+        )?;
+
+        let mut x_a_cell = x_a.cell();
+        let mut x_a = x_a.value();
+        let mut y_a = y_a;
+
+        let generators: Vec<Option<(pallas::Base, pallas::Base)>> =
+            if let Some(generators) = generators {
+                generators.into_iter().map(Some).collect()
+            } else {
+                vec![None; piece.num_words()]
+            };
+
+        for (row, gen) in generators.iter().enumerate() {
+            let x_p = gen.map(|gen| gen.0);
+            let y_p = gen.map(|gen| gen.1);
+
+            // Assign `x_p`
+            region.assign_advice(
+                || "x_p",
+                config.x_p,
+                offset + row,
+                || x_p.ok_or(Error::SynthesisError),
+            )?;
+
+            // Compute and assign `lambda1, lambda2`
+            let lambda1 = x_a
+                .zip(y_a)
+                .zip(x_p)
+                .zip(y_p)
+                .map(|(((x_a, y_a), x_p), y_p)| (y_a - y_p) * (x_a - x_p).invert().unwrap());
```

Comment:
These inversions would be batchable if https://github.com/zcash/halo2/pull/293 merges.